### ISSUE TITLE
disable CMA in vader

### ISSUE THREAD

**Issue by hunsa** [open-mpi/ompi#3270]
Hello all,

We experienced problems with Open MPI when communication larger
messages, for example with `MPI_Gather`, `MPI_Allgather`, etc.  Large
means messages larger than 1k (for very small messages the problem did
not occur).

We received error messages like this
  Read -1, expected 8000, errno = 38

I suspected that there is something wrong with the CMA support.

And indeed, until Open MPI 2.0.2 the configure script would
have the following result
  `#define OPAL_BTL_SM_HAVE_CMA 0`

Now, the new CMA detection method in 2.1.0 leads to 
   `#define OPAL_BTL_SM_HAVE_CMA 1`

That wouldn't be a problem if we could disable CMA in vader during
configure, but it does not seem to be possible.  So, `--with-cma=no` or
`--without-cma` will have no effect and we will end up with
   `#define OPAL_BTL_SM_HAVE_CMA 1`

Currently, we can set 
  `OMPI_MCA_btl_vader_single_copy_mechanism=none`
and the error messages will not show up.

It would be great if we could manually disable CMA support during
configure.

Thank you

**Comment by ggouaillardet** [comment]
Meanwhile, can you try
```
opal_check_cma_happy=0 configure ...
make install
```
and see whether it helps ?

fwiw, we currently only use --with-cma to abort if cma support cannot be built.
I will fix that so we also support --without-cma

**Comment by hunsa** [comment]
`opal_check_cma_happy=0 configure` works and it is much better than having to set `OMPI_MCA_btl_vader_single_copy_mechanism` each time.

**Comment by hjelmn** [comment]
Vader doesn't use BTL_SM_ variables. Look for OPAL_BTL_VADER_HAVE_XPMEM. 

Looks to me like CMA support either doesn't exist on your platform or we have the wrong syscall numbers for the platform. That errno translates to function not implemented. What is the kernel version and platform. It would also be helpful to see your config.log for the build.

**Comment by hunsa** [comment]
@hjelmn Your are right. But the CMA detection macro (`OPAL_CHECK_CMA`) is the same for SM and VADER. Thus, if CMA is enabled, it will be enabled in both cases. It is as you say: our platform has no CMA support. So, I would like to disable it.
Here is the `config.log` (IP/host information filtered out) for a default configure run.

[config.log.zip](https://github.com/open-mpi/ompi/files/890660/config.log.zip)

**Comment by hjelmn** [comment]
Hmm, --with-cma=no was working at some point. Looks like the check is missing. Also odd that the CMA test passed. Are you cross-compiling?

**Comment by hunsa** [comment]
No, we are not cross-compiling. I have been compiling several Open MPI versions with the same configure options for some time, and the problem with CMA only showed up with 2.1.0.

**Comment by hjelmn** [comment]
We started enabling CMA by default in v2.1.x. We have a configure check that calls process_vm_writev to see if the call is supported. That is why I ask if you are cross-compiling (this includes different kernel versions/configurations BTW).

**Comment by hjelmn** [comment]
Can you try --with-cma=no with #3272 and verify that it works for you.

**Comment by ggouaillardet** [comment]
@hjelmn i just realized you issued a similar PR a few hours ago ...
i tested ```--without-cma``` and it produces the expected effect
```
$ grep -R CMA opal/include | grep define
opal/include/opal_config.h:#define HAVE_RDMA_RDMA_CMA_H 1
opal/include/opal_config.h:#define OPAL_BTL_SM_HAVE_CMA 0
opal/include/opal_config.h:#define OPAL_BTL_VADER_HAVE_CMA 0
opal/include/opal_config.h:/* Need CMA syscalls defined */
```
(on my centos7 box, ```...?HAVE_CMA``` is ```1``` by default)

feel free to merge either #3272 or #3275 and close the unused one.
